clusterctl/cmd/validate_cluster.go

```diff
@@ -17,26 +17,130 @@ limitations under the License.
 package cmd
 
 import (
+	"fmt"
+
 	"github.com/golang/glog"
 	"github.com/spf13/cobra"
-	"sigs.k8s.io/cluster-api/errors"
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/common"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
 )
 
+type ValidateClusterOptions struct {
+	ClusterName string
+	Namespace   string
+}
+
+var vco = &ValidateClusterOptions{}
+
 var validateClusterCmd = &cobra.Command{
 	Use:   "cluster",
 	Short: "Validate a cluster created by cluster API.",
 	Long:  `Validate a cluster created by cluster API.`,
 	Run: func(cmd *cobra.Command, args []string) {
-		if err := RunValidateCluster(); err != nil {
-			glog.Exit(err)
-		}
+		RunValidateCluster()
 	},
 }
 
 func init() {
+	validateClusterCmd.Flags().StringVarP(
+		&vco.ClusterName, "cluster-name", "", "",
+		"The name of the cluster to be validated. This can be omitted if there is only one cluster being managed in the given namespace.")
+	validateClusterCmd.Flags().StringVarP(
+		&vco.Namespace, "namespace", "n", v1.NamespaceDefault,
+		"If present, validate the cluster in the specified namespace.")
 	validateCmd.AddCommand(validateClusterCmd)
 }
 
-func RunValidateCluster() error {
-	return errors.NotImplementedError
+func RunValidateCluster() {
+	// TODO(wangzhen127): Also validate the cluster in addition to the cluster API objects.
+	cluster := validateClusterAPIObjects()
+
+	glog.Infof("Pass. The cluster %q in namespace %q is validated.", cluster.Name, vco.Namespace)
+}
+
+func validateClusterAPIObjects() *v1alpha1.Cluster {
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		&clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		glog.Fatalf("Failed to create client for talking to the apiserver: %v", err)
```

Comment:
Recommend that this function adds a second return value of 'error' and the calling function, calls glog.Fatal(...) like this:

```
cluster, error := validateClusterAPIObjects()
if err != nil {
    glog.Fatal(err)
}
```

It's generally undesirable to have calls to exit an application littered in various places because it becomes more complicated to reason about application behavior as there are more possible ways to exit. It's harder to standardize the error messages that are displayed in these exit scenarios as well as there is not a single location where the final error message is printed and exit is called. In addition, it is more difficult to unit test code that can call exit. 

And like I said in a previous comment, ideally we would keep the single glog.Exit(...) call in the main Run function defined on the command to match all other commands in clusterctl. That will be easier for future developers to understand.

Reply:
Makes sense. Updated